File: content/problems/1298-maximum-candies-you-can-get-from-boxes/solutions/owned-key-open-queue/Solution.java
class Solution {
    public int maxCandies(int[] status,int[] candies,int[][] keys,int[][] containedBoxes,int[] initialBoxes){
        int n=status.length,total=0;
        boolean[] owned=new boolean[n],hasKey=new boolean[n],opened=new boolean[n];
        Queue<Integer> queue=new ArrayDeque<>();
        for(int box:initialBoxes){
            owned[box]=true;
            queue.offer(box);
        }while(!queue.isEmpty()){
            int box=queue.poll();
            if(opened[box]||!owned[box]||status[box]==0&&!hasKey[box])continue;
            opened[box]=true;
            total+=candies[box];
            for(int key:keys[box]){
                hasKey[key]=true;
                if(owned[key])queue.offer(key);
            }
            for(int child:containedBoxes[box]){
                owned[child]=true;
                queue.offer(child);
            }
        }
        return total;
    }
}

File: content/problems/1298-maximum-candies-you-can-get-from-boxes/solutions/owned-key-open-queue/Solution.kt
class Solution {
    fun maxCandies(status:IntArray,candies:IntArray,keys:Array<IntArray>,containedBoxes:Array<IntArray>,initialBoxes:IntArray):Int{
        val n=status.size;
        val owned=BooleanArray(n);
        val hasKey=BooleanArray(n);
        val opened=BooleanArray(n);
        val queue=java.util.ArrayDeque<Int>();
        for(box in initialBoxes){
            owned[box]=true;
            queue.addLast(box)
        };
        var total=0;
        while(queue.isNotEmpty()){
            val box=queue.removeFirst();
            if(opened[box]||!owned[box]||status[box]==0&&!hasKey[box])continue;
            opened[box]=true;
            total+=candies[box];
            for(key in keys[box]){
                hasKey[key]=true;
                if(owned[key])queue.addLast(key)
            };
            for(child in containedBoxes[box]){
                owned[child]=true;
                queue.addLast(child)
            }
        };
        return total
    }
}
